src/libsystemd/sd-device/sd-device.c

```diff
@@ -1873,13 +1872,33 @@ _public_ int sd_device_set_sysattr_value(sd_device *device, const char *sysattr,
                 return 0;
         }
 
+        value_len = strlen(_value);
+
+        /* drop trailing newlines */
+        while (value_len > 0 && _value[value_len - 1] == '\n')
+                _value[--value_len] = '\0';
+
+        /* value length is limited to 4k */
+        if (value_len > 4096)
+                return -EINVAL;
+
+        value = strdup(_value);
+        if (!value)
+                return -ENOMEM;
+
         r = sd_device_get_syspath(device, &syspath);
         if (r < 0)
                 return r;
 
         path = strjoina(syspath, "/", sysattr);
-        r = lstat(path, &statbuf);
-        if (r < 0) {
+
+        fd = open(path, O_WRONLY | O_CLOEXEC | O_NOFOLLOW);
+        if (fd < 0) {
+                if (errno == ELOOP)
+                        return -EINVAL;
+                if (errno == EISDIR)
+                        return -EISDIR;
+
                 value = strdup("");
```

Comment:
Beats me ;) I think it's very dubious. I think the idea was that if the attribute does not exist, the internal state should be set to empty. I'm not sure what this function is used for, and I was trying not to change the semantics.